Convert project to `poetry`

guidopetri opened the pull request:
## What type of PR is this? 

- [x] Other

## Description

This PR converts us from `requirements.txt`-pinned dependencies to `poetry`, which has much more stable dependency chains and should prevent us from having updated dependencies break our project.

I added two groups (besides the main group): `dev`, for development dependencies, and `all_ds`, to keep the separation we had with the requirements files.

## How is this tested?

- [x] Unit tests (pytest, jest)
- [x] E2E Tests (Cypress)
- [x] Manually

Untested as of yet (so don't merge), but the plan is to:

- build the image
- `pip freeze > image_packages.txt`
- on a new installation, install dependencies with `poetry install`
- `pip freeze > poetry_packages.txt`
- `diff image_packages.txt poetry_packages.txt`

## Related Tickets & Documents

#5910 

## Mobile & Desktop Screenshots/Recordings (if there are UI changes)

N/A

justinclift commented:
No objection here, it sounds like it'll be useful.

@getredash/maintainers @arikfr This ok with yourselves? :smile:

guidopetri commented:
This isn't quite ready yet :P

snickerjp commented:
I thought Poetry was a good attempt.

justinclift commented:
> This isn't quite ready yet :P

No worries.  Was more thinking we should ensure the concept itself is ok with people. :smile:

guidopetri commented:
#6428

guidopetri commented:
`grep -Ir requirements .` returns only hits in `./bin/upgrade` now, and I can't find any programmatic use for that script. @arikfr , do you remember what this was used for? Should we update it?

I've made one important change to the Dockerfile: I removed the `skip_ds_deps`, `skip_dev_deps`, `test_all_deps` build args and replaced them with an `install_groups` build argument. This is passed directly to poetry and can control which dependency groups are installed. By default, it will install `main`, `all_ds` and `dev`; I changed the Cypress yaml and CI workflows accordingly. The image builds on my local now with `docker build -t redash .`.

I'm now going to create a venv to compare the poetry install with the requirements.txt install and see if there are any differing packages.

guidopetri commented:
Ok, here's the diff list I got:

```
55a56
> futures==3.0.5
74c75
< inflection==0.5.1
---
> inflection==0.3.1
115a117
> pip==23.2.1
159c161,162
< rdflib==7.0.0
---
> rdflib==6.3.2
> redash==23.9.0       /home/sid/Git/redash
175a179
> setuptools==68.2.0
210a215
> wheel==0.40.0
```

wheel/setuptools are from my pyenv, and redash is because I installed without `--no-root`. The `inflection` version being different is interesting, as well as `rdflib`. I suspect this is some incompatibility in one of the packages with the newer versions of these two packages. As long as the tests pass, I'm good merging. @getredash/maintainers , any objections to this?

justinclift reviewed (changes requested):
Just noticed that this is missing the dependency updates from the rq related PR yesterday: https://github.com/getredash/redash/commit/41495ba940ac9d1bb2beef3456b89b5742f11eec

They should be added yeah?

justinclift commented on pyproject.toml:
```diff
@@ -16,3 +16,152 @@ multi_line_output = 3
+
+[tool.poetry]
+name = "redash"
+version = "23.09.0"
```
Does poetry not understand things like `-dev` in version identifiers?

guidopetri replied:
Resolved!

guidopetri commented:
d727602 should be the rq changes. Is it not showing up on the diff?

I forgot about the `dev` suffix, I think poetry is yine woth it but let me add that and double check. It looks like I also missed something in the CI.

justinclift commented:
> Is it not showing up on the diff?

Ahhh, yeah that wasn't showing up for me.  No worries. :smile:

guidopetri commented:
Alright, I've enabled auto-merge on this PR. @justinclift when you have a chance, can you re-review to merge?

arikfr commented:
`bin/upgrade` is a relic from the pre-docker days. We should just delete it.

arikfr commented on pyproject.toml:
```diff
@@ -16,3 +16,151 @@ multi_line_output = 3
+name = "redash"
+version = "23.09.0-dev"
+description = "Make Your Company Data Driven. Connect to any data source, easily visualize, dashboard and share your data."
+authors = ["Your Name <you@example.com>"]
```
```suggestion
authors = ["Redash maintainers and contributors"]
```

Or does it have to be an email address?

guidopetri commented:
Shoot, the auto-merge didn't quite help here did it. Sorry @arikfr ! I'll create a followup PR. The author does indeed have to include an email address; do we have one on the `redash.io` domain that we could use?

arikfr commented:
I can set something up on redash.io - just wonder where to send it? Maybe there is some email->discord integration?

justinclift commented:
> Maybe there is some email->discord integration?

Interesting idea.  We'd better make sure it never sends anything confidential via that channel though (password resets, etc), unless we're sure it can direct to some private for-admins-only channel. :smile:
